scylla/src/transport/node.rs

```diff
@@ -267,27 +269,53 @@ pub(crate) struct ResolvedContactPoint {
     pub(crate) datacenter: Option<String>,
 }
 
+#[derive(Error, Debug)]
+pub(crate) enum DnsLookupError {
+    #[error("Failed to perform DNS lookup within {0}ms")]
+    Timeout(u128),
+    #[error("Empty address list returned by DNS for {0}")]
+    EmptyAddressListForHost(String),
+    #[error(transparent)]
+    IoError(#[from] io::Error),
+}
+
+/// Performs a DNS lookup with provided optional timeout.
+async fn lookup_host_with_timeout<T: ToSocketAddrs>(
+    host: T,
+    hostname_resolution_timeout: Option<Duration>,
+) -> Result<impl Iterator<Item = SocketAddr>, DnsLookupError> {
+    if let Some(timeout) = hostname_resolution_timeout {
+        match tokio::time::timeout(timeout, lookup_host(host)).await {
+            Ok(res) => res.map_err(Into::into),
+            // Elapsed error from tokio library does not provide any context.
+            Err(_) => Err(DnsLookupError::Timeout(timeout.as_millis())),
+        }
+    } else {
+        lookup_host(host).await.map_err(Into::into)
+    }
+}
+
 // Resolve the given hostname using a DNS lookup if necessary.
 // The resolution may return multiple IPs and the function returns one of them.
 // It prefers to return IPv4s first, and only if there are none, IPv6s.
-pub(crate) async fn resolve_hostname(hostname: &str) -> Result<SocketAddr, io::Error> {
-    let addrs = match lookup_host(hostname).await {
+pub(crate) async fn resolve_hostname(
+    hostname: &str,
+    hostname_resolution_timeout: Option<Duration>,
+) -> Result<SocketAddr, DnsLookupError> {
+    let addrs = match lookup_host_with_timeout(hostname, hostname_resolution_timeout).await {
         Ok(addrs) => itertools::Either::Left(addrs),
         // Use a default port in case of error, but propagate the original error on failure
         Err(e) => {
-            let addrs = lookup_host((hostname, 9042)).await.or(Err(e))?;
+            let addrs = lookup_host_with_timeout((hostname, 9042), hostname_resolution_timeout)
+                .await
+                .or(Err(e))?;
             itertools::Either::Right(addrs)
         }
```

Comment:
This means that the timeout is effectively twice the value provided in the config...
I'm not convinced it's correct. This must be at least noted in documentation.

Reply:
Hmm... you are right. I wonder whether we should leave it as is, or rather recompute the timeout for the second try in case first failed.

```rust
    let deadline = hostname_resolution_timeout.map(|dur| tokio::time::Instant::now() + dur);
    let addrs = match lookup_host_with_timeout(hostname, hostname_resolution_timeout).await {
        Ok(addrs) => itertools::Either::Left(addrs),
        // Use a default port in case of error, but propagate the original error on failure
        Err(e) => {
            let new_timeout =
                deadline.map(|deadline| deadline.duration_since(tokio::time::Instant::now()));
            let addrs = lookup_host_with_timeout((hostname, 9042), new_timeout)
                .await
                .or(Err(e))?;
            itertools::Either::Right(addrs)
        }
    };
```

Reply:
If the first attempt failed from reasons different that a timeout, then such recomputation makes perfect sense. However, if it failed due to a timeout (because the wrong port somehow caused a timeout - idk how likely this scenario is), then we won't be able to do the second attempt.

Reply:
Any ideas @Lorak-mmk?

Reply:
We could also execute `resolve_hostname` function with a given timeout (i.e. `tokio::time::timeout(t, resolve_hostname)`). It's much easier and cleaner than recomputing the timeout. However, in such case it's still true that if first attempt takes too long, we won't be able to do a second attempt.

Reply:
When can it happen that the port is not specified and what is the expected behavior of `lookup_host` in this case?

If it always fails immediately with a specific error, then we can check this error and perform lookup with default port with the same timeout.

If it can timeout, or take some time in general, then I'd perform another lookup with the same timeout and document it.

Reply:
From my investigation, it looks like port is not used in the actual DNS lookup (which makes sense).

What's the difference between `String` and `(String, u16)`, then?
When `String` is provided, both `tokio` and `std::net` expect that it is of form `"host:port"`. If it's not the case, the error is returned immidiately. Otherwise, `host` part is used in DNS lookup which may timeout.

When `(String, u16)` is provided, the DNS lookup is performed based on the string - port is ignored during lookup (same as above).

This means that:
- if hostname is of form `"addr:port"`, then the first lookup timeouts **iff** the second lookup (with default port) timeouts
- if hostname is not of this form, then the first lookup will fail immidiately. We can try with default port.

In that case, I think that we can match the error returned from `lookup_host_with_timeout` If it's something other than `DnsLookupError::Timeout(_)`, then we can try again with the same timeout.

Reply:
Ok. Let's describe this behavior in the docs.